doc/music.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH MUSIC 1 "March  5, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
music \- launch a multi-simulation using MUSIC
.SH SYNOPSIS
.B mpirun
.RI [ mpirun-options ]
.B  music
.RI [ options ] " config"
.br
.B music
.RI [ options ] " config" ...
.SH DESCRIPTION
This manual page documents briefly the
.B music
utility.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBmusic\fP is used to launch a multi-simulation, i.e., an MPI job
partitioned between a set of parallel programs.  Such programs need to
be MUSIC-aware (see documentation for the MUSIC library).  MUSIC also
provides facilities for these programs to exchange data.
.PP
The second form of the command above (without \fBmpirun\fP) can be
used to print out an application rank map or version information.
.SH OPTIONS
The
.B music
utility follows the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the MUSIC manual.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-m, \-\-map
Print application rank map.
.TP
.B \-v, \-\-version
Show version of program.
.\" .SH SEE ALSO
.\" .BR bar (1),
.\" .BR baz (1).
.\" .br
.\" The programs are documented fully by
.\" .IR "The Rise and Fall of a Fooish Bar" ,
.\" available via the Info system.
.SH AUTHOR
MUSIC was written by Mikael Djurfeldt and Örjan Ekeberg for INCF.
.PP
This manual page was written by Mikael Djurfeldt <mdj@debian.org>,
for the Debian project (but may be used by others).


NEWS
MUSIC NEWS --- history of user-visible changes.
Copyright (C) 2009, 2010 INCF

Please send MUSIC bug reports to music-bugs@incf.org.


Changes in 1.0.7:

* music library

** Bug fix release.


Changes in 1.0.6:

* music library

** Fixes for GCC version 4.4.

** Don't include RPATH in libmusic.so.


Changes in 1.0.5:

* music library

** Optimizations of event routing.  MUSIC now discovers ranges of
   indices which can be handled collectively by one entry in the event
   router, even in the case that this is not apparent from index maps
   but due to the effect of negotiations with the port connected on
   the other side.

** Bug fixes.
   

Changes in 1.0.4:

* test directory

** New example: messages.music

* music library

** Message ports now work correctly.

** Optimizations.


Changes in 1.0.3:

* music library

** Fixes a bug which occurs in Runtime::finalize () when using the
   threading version of the Setup constructor under OpenMPI 1.2.7.

** Other smaller fixes.


Changes in 1.0.2:

* music library

** Better error reporting.

** Bug fixes.

** New README file in test directory.

** New demo in test directory.


Changes in 1.0.1:

* music utility

** Bug fixes

** Improved error reporting

** New option '-m' prints application rank mapping

** New option '-v' prints version information

** Debian packaging support


First release: 1.0

Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


utils/music.cc
/*
 *  This file is part of MUSIC.
 *  Copyright (C) 2007, 2008, 2009 INCF
 *
 *  MUSIC is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  MUSIC is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <string>
#include <fstream>

#include "config.h"

#include <music/version.hh>

#include "music/application_mapper.hh"
#include "mpidep/mpidep.hh"

extern "C" {
#include <sys/types.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
}
using std::string;

void
usage (int rank)
{
  if (rank <= 0)
    {
      std::cout << "Usage: mpirun ... music [OPTION...] CONFIG" << std::endl
		<< "`music' launches an application as part of a multi-simulation job." << std::endl << std::endl
		<< "  -h, --help            print this help message" << std::endl
		<< "  -m, --map             print application rank map" << std::endl
//		<< "  -f, --file-map        creates a file with the list of environment variable of each application" << std::endl
//		<< "  -e, --export-scripts  export launcher scripts" << std::endl
		<< "  -v, --version         prints version of MUSIC library" << std::endl
		<< std::endl
		<< "Report bugs to <music-bugs@incf.org>." << std::endl;
    }
  exit (0);
}


void
print_map (MUSIC::Configuration* config)
{
  MUSIC::ApplicationMap* a = config->applications ();
  std::cout << "rank\tapplication" << std::endl;
  for (MUSIC::ApplicationMap::iterator i = a->begin (); i != a->end (); ++i)
    {
      int first = i->leader ();
      int nProc = i->nProc ();
      std::cout << first;
      if (nProc > 1)
	std::cout << "-" << first + nProc - 1;
      std::cout << '\t' << i->name () << std::endl;
    }
}

/*const static char* const configEnvVarName = "_MUSIC_CONFIG_";
const static char* const mapFName = "music.map";
void
export_scripts (MUSIC::ApplicationMapper* map)
{
  MUSIC::Configuration* config = map->config ();
  MUSIC::ApplicationMap* a = config->applications ();
  for (MUSIC::ApplicationMap::iterator i = a->begin (); i != a->end (); ++i)
    {
      std::string name = i->name ();
      std::string fname = name + ".sh";
      std::ofstream script (fname.c_str ());
      script << "#!/bin/sh" << std::endl << std::endl;
      map->mapConnectivity (name);
      MUSIC::Configuration* config = map->config (i->color());
      config->writeEnv ();
      script << "export " << configEnvVarName
	     << "=\"" << getenv (configEnvVarName)
	     << '"' << std::endl;
      std::string binary;
      config->lookup ("binary", &binary);
      std::string args;
      config->lookup ("args", &args);
      script << binary << ' ' << args << std::endl;
      script.close ();
    }
}

void export_map(MUSIC::ApplicationMapper* map)
{
	MUSIC::Configuration* config = map->config ();
	MUSIC::ApplicationMap* a = config->applications ();
    std::ofstream map_file (mapFName);
	  for (MUSIC::ApplicationMap::iterator i = a->begin (); i != a->end (); ++i)
	    {
		  map->mapConnectivity (i->name());
		  MUSIC::Configuration* config = map->config (i->color());
		  config->writeEnv ();
		  map_file <<  getenv (configEnvVarName) << std::endl;
	    }
	  map_file.close ();
}*/


void
print_version (int rank)
{
  if (rank <= 0)
    {
      std::cout << "MUSIC " << MUSIC::version () << std::endl
		<< "Copyright (C) 2007-2012 INCF." << std::endl
		<< "You may redistribute copies of MUSIC" << std::endl
		<< "under the terms of the GNU General Public License." << std::endl
		<< "For more information about these matters, see the file named COPYING." << std::endl;
    }
  exit (0);
}


void
launch (MUSIC::Configuration* config, char** argv)
{
  string binary;
  config->lookup ("binary", &binary);
  config->writeEnv ();
  string wd;
  if (config->lookup ("wd", &wd))
    {
      if (chdir (wd.c_str ()))
	goto error_exit;
    }
  execvp (binary.c_str (), argv);

 error_exit:
  // if we get here, something is wrong
  std::cerr << "Error during launching of binary " << binary << std::endl;
  perror ("MUSIC");
  exit (1);
}

/* remedius
 * option -f to the music launcher tool is added.  This option causes the
 *	music launcher tool to write a file  that contains mapping information of the rank (rank range)
 *	to the according launching environment variable.
 *	This file can be used while launching multiple applications in MPMD mode.
 */
int
main (int argc, char *argv[])
{
  // predict the rank MPI::Init will give us using
  // mpi implementation dependent code from ../mpidep
  int rank = -1;
#if MUSIC_USE_MPI
  rank = getRank (argc, argv);
#endif

  bool do_print_map = false;
  bool do_export_scripts = false;
  bool do_export_map = false;
  
  opterr = 0; // handle errors ourselves
  while (1)
    {
      static struct option longOptions[] =
	{
	  {"help",           no_argument,       0, 'h'},
	  {"map",            required_argument, 0, 'm'},
//	  {"export-scripts", required_argument,       0, 'e'},
//	  {"file-map",        required_argument,       0, 'f'},
	  {"version",        no_argument,       0, 'v'},
	  {0, 0, 0, 0}
	};
      /* `getopt_long' stores the option index here. */
      int option_index = 0;

      // the + below tells getopt_long not to reorder argv
      int c = getopt_long (argc, argv, "+hm:v", longOptions, &option_index);
      /* detect the end of the options */
      if (c == -1)
	break;
      switch (c)
	{
	case '?':
	  break; // ignore unknown options
	case 'h':
	  usage (rank);
	case 'm':
	  do_print_map = true;
	  continue;
/*	case 'e':
	  do_export_scripts = true;
	  continue;
	case 'f':
		do_export_map = true;
	  continue;*/
	case 'v':
	  print_version (rank);

	default:
	  abort ();
	}
    }

  // extract the configuration file name using
  // mpi implementation dependent code from ../mpidep
  std::istream* configFile = getConfig (rank, argc, argv);

  if (!*configFile)
    {
      if (rank <= 0)
	std::cerr << "MUSIC: Couldn't open configuration file "
		  << argv[1] << std::endl;
      exit (1);
    }

  MUSIC::Configuration config;
  MUSIC::ApplicationMapper mapper (&config);
  mapper.map (configFile, rank);
  if (do_print_map)
    {
      if (rank <= 0)
	print_map (&config);
    }

  /*  if (do_export_scripts)
    {
      if (rank <= 0)
	export_scripts (&map);
    }
  if(do_export_map)
  {
	  if (rank <=0)
		  export_map(&map);

  }*/
  if (do_print_map) //|| do_export_scripts || do_export_map)
    return 0;
  
  if (rank == -1)
    {
      std::cerr << "MUSIC: Unable to determine process rank." << std::endl
		<< "MUSIC: Did you launch music using mpirun?" << std::endl
		<< "MUSIC: If so, check the comments about porting in README." << std::endl;
      exit (1);
    }

  launch (&config, argv);

  return 0;
}
